src/util.rs

```diff
@@ -0,0 +1,15 @@
+/// Utility functions.
+
+/// Returns true if a slice is a permutation.
+///
+/// That is, all the elements in `0..len` occur exactly once in the slice.
+pub fn is_permutation(v: &[usize]) -> bool {
+    let n = v.len();
+    let mut seen = (0..n).map(|_| false).collect::<Vec<_>>();
```

Comment:
I think this could just be

```suggestion
    let mut seen = vec![false; n];
```